google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java

```diff
@@ -474,19 +470,17 @@ public void start(Listener<RespT> responseListener, Metadata metadata) {
 
   private Metadata newMetadata(String resource) {
     Metadata metadata = new Metadata();
-    metadata.put(RESOURCE_PREFIX_KEY, extractHeader(resource));
-    metadata.put(API_CLIENT_KEY, xGoogApiClientHeader);
-    return metadata;
-  }
 
-  @VisibleForTesting
-  String extractHeader(String resource) {
-    Matcher m = DATABASE_PATTERN.matcher(resource);
-    if (m.matches()) {
-      return m.group("database");
+    Map<Metadata.Key<String>, String> headers = headerProvider.getHeadersAsMetadata();
```

Comment:
I think it might be cleaner if this whole logic is put inside SpannerHeaderProvider. It can have a method like
void populateHeaders(Metadata metadata, String resource, String project);

Reply:
(same comment as above): Converted `SpannerHeaderProvider` to `SpannerMetadataProvider`, which now has only one method `newMetadata()`, with everything encapsulated there.